### ISSUE TITLE
Need information on working of kernel.json file

### ISSUE THREAD

**Issue by aniket02k** [jupyter/enterprise_gateway#884]
Hi,
We wanted to use customized kernel images for different users. Users will use the same enterprise gateway instance. But, the kernel images might be different for each user. Wanted to know how the kernel.json file is invoked when any kernel is launched.

Thanks,
Aniket

**Comment by kevin-bates** [comment]
Hi Aniket - thanks for your question. I'm going to assume you're referring to running EG in either Kubernetes or a Docker environment, and my response is generally the same for each.

There are two basic approaches that can be implemented for allowing different users to have different kernel images for similarly configured kernels otherwise:
1. Utilize the `KERNEL_IMAGE` environment variable and _flow it_ from the client.  All client-side environment variables prefixed with `KERNEL_` will be made available to the underlying kernel image.  [Some, like `KERNEL_IMAGE`, are recognized by EG](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/config-options.html#per-kernel-environment-overrides) and, in this case, EG will substitute the image it finds in the `kernel.json` file with the value of `KERNEL_IMAGE`.   As a result, you would simply have each of your users set the appropriate value for `KERNEL_IMAGE` prior to their launch of Notebook or JupyterLab when configured to use the `--gateway-url`.
2. You could create separate kernelspecs for each different image.  Different users would then select their image.  With this approach you could use the [`authorized_users` functionality](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/config-options.html#per-kernel-configuration-overrides) and only allow a particular kernelspec to be "visible/usable" by the specific users.

If each user's environment is identical to each other, then approach 1 is probably preferred (and easier regardless).  However, if you want different users using differently configured environments - Spark vs. non-Spark or some with Scala, others with R - then approach 2 would be used.

Also keep in mind that with approach 1, the same image would be used regardless of which kernelspec was selected as the existence of a `KERNEL_IMAGE` value supersedes the image name embedded in the `kernel.json` file.  So I suppose its only viable if a given user is "pinned" to a specific kernel specification or similar specifications where the same underlying image can be used.

Can you elaborate on how the various users would be differentiated based on the above?  Also keep in mind that EG only supports Python, R and Scala kernels.

As for "how the kernel.json file is invoked when any kernel is launched", this is a larger question and depends on the container environment and (for now) if Spark is used or not.  But, generally speaking, the server essentially invokes the `argv` stanza that's in the `kernel.json` file.  In the case of EG, that can do quite a lot.  On Kubernetes, it essentially builds a yaml file that becomes the kernel POD configuration. This yaml file is derived from a template that has various values substituted - both system-defined values as well as the various `KERNEL_` values.  However, for Kubernetes on Spark, that launch goes through Spark-submit instead and we can't use the templated approach, so you see many variables set in the SPARK OPTIONS portion of the `kernel.json`.  (In Spark 3.0, the driver and executor pods can be configured using a yaml file and we [plan to support that](#559) in our EG 3.0 release.)

At any rate, let's leave things here for now.  If I didn't answer your question or you have additional questions, please feel free to ask here and we can continue the discussion.

**Comment by aniket02k** [comment]
> Can you elaborate on how the various users would be differentiated based on the above? Also keep in mind that EG only supports Python, R and Scala kernels.


Hi Kevin, 
Thanks for your inputs.
Each user will have separate jupyterlab client and image configurations will be passed to eg from lab client.

Regards,
Aniket

**Comment by aniket02k** [comment]
Hi Kevin,
I am passing `KERNEL_IMAGE` env from client to eg. Still its referring to  image which is specified in `kernel.json` file.  Any pointers on what is being missed?

**Comment by kevin-bates** [comment]
Shoot - I must apologize Aniket. When I grepped for `KERNEL_IMAGE` I saw that it was being used to override the image name in the process_proxy stanza of the `kernel.json` file, but I didn't realize that it is getting pulled from EG's environment and not from the environment coming from the client.  I'm really sorry to mislead you about that.  `KERNEL_IMAGE` is intended to be a general EG override.  In thinking more about this, I believe this is a bug.  That name should be `EG_KERNEL_IMAGE` and should be used as an overall fallback when both `KERNEL_IMAGE` and the process_proxy do not define values.  I have opened issue #886 to address this.

Until the issue is resolved, I think you'll need to take the multi-kernelspec approach where each kernelspec identifies the user.  To apply appropriate filtering, so that a given user only sees their kernel, I recommend the following:
1. Set env `KERNEL_USERNAME` on each client representing the user.  This _does_ flow and I'm assuming you're already doing this.
2. In each of the per-user kernelspecs, update the `image_name` to the appropriate image for that user.  If some users will share the same image, they can be grouped via the same kernelspec directory.
3. In each of the per-user kernelspecs, add an `authorized_users` list containing the appropriate user names.  These values will match the value of the `KERNEL_USERNAME` env (case-insensitive).  Again, if multiple users are sharing the same image/kernelspec, they should be listed in this entry.  This is the vehicle by which the kernels displayed to the user are filtered.

Here are a couple of examples, one for a single-user specification, the other for a group of users:

### Single-user specification
This entry would be located in `/usr/local/share/jupyter/kernels/alice` and user 'alice' will see 'Python on Kubernetes' (i.e., the `display_name` value) as their kernel to use.
```JSON
{
  "language": "python",
  "display_name": "Python on Kubernetes",
  "metadata": {
    "process_proxy": {
      "class_name": "enterprise_gateway.services.processproxies.k8s.KubernetesProcessProxy",
      "config": {
        "image_name": "image_name_assigned_to_alice:tag",
        "authorized_users": "alice"
      }
    }
  },
  "env": {
  },
  "argv": [
    "python",
    "/usr/local/share/jupyter/kernels/alice/scripts/launch_kubernetes.py",
    "--RemoteProcessProxy.kernel-id",
    "{kernel_id}",
    "--RemoteProcessProxy.response-address",
    "{response_address}",
    "--RemoteProcessProxy.public-key",
    "{public_key}"
  ]
}
```
### Group-based specification
This entry would be located in `/usr/local/share/jupyter/kernels/eds` and users 'ed', 'ned', and 'ted' will see 'Python on Kubernetes' (i.e., the `display_name` value) as their kernel to use.
```JSON
{
  "language": "python",
  "display_name": "Python on Kubernetes",
  "metadata": {
    "process_proxy": {
      "class_name": "enterprise_gateway.services.processproxies.k8s.KubernetesProcessProxy",
      "config": {
        "image_name": "image_name_assigned_to_eds:tag",
        "authorized_users": "ed, ned, ted"
      }
    }
  },
  "env": {
  },
  "argv": [
    "python",
    "/usr/local/share/jupyter/kernels/eds/scripts/launch_kubernetes.py",
    "--RemoteProcessProxy.kernel-id",
    "{kernel_id}",
    "--RemoteProcessProxy.response-address",
    "{response_address}",
    "--RemoteProcessProxy.public-key",
    "{public_key}"
  ]
}
```
Note the path to the `launch_kubernetes.py` script in each `<user>/scripts` directory. Rather than duplicate the entire kernelspec directory for each user/group, you could keep a primary kernelspec directory around (e.g., `python_kubernetes`) and reference _that_ script in the `argv` stanza.  Then you'd only need to replicate the `kernel.json` (and logo image) for each user/group entry.  However, to prevent that primary kernelspec from showing up in the various user's selection list, it should have either an `authorized_users` list of users exclusive to the union of the other kernels, or use an `unauthorized_users` list consisting of the union of the other others.

If an admin needs access to each of the various kernels, then you might want to differentiate the kernelspec `display_name` values, otherwise, you'll see duplicates in your selection list.  However, once the issue is resolved and `KERNEL_IMAGE` can flow, all users will see the same kernel name again.

I'm really sorry to mislead you.  `KERNEL_IMAGE` just isn't something that is used often and isn't generally useful because it pins the kernel image across all kernelspecs.

**Comment by kevin-bates** [comment]
Hi @aniket02k - I have [pushed an enterprise-gateway image](https://hub.docker.com/r/elyra/enterprise-gateway/tags) that includes the fix.  Please deploy `elyra/enterprise-gateway:886` (the tag is the issue number) and you should find that `KERNEL_IMAGE` flows from the client.

Also, if you're building your own kernel images, please derive those images `FROM` an existing image, or be sure to include the appropriate bootstrapping software in your images.  See [Custom Kernel Images](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/docker.html#custom-kernel-images) for these kinds of details.  Thanks.

If you could confirm you see the expected behavior, I will move forward with creating a pull request, etc.

**Comment by aniket02k** [comment]
Hi Kevin, 
Can you please provide the fix (code changes), so that we can integrate the fix with derived image and test the same?

Regards,
Aniket

**Comment by kevin-bates** [comment]
Thanks Aniket - please see https://github.com/jupyter/enterprise_gateway/pull/889 and let us know how it goes.  We'd like to get this into the next release (2.3) which is imminent.

**Comment by kevin-bates** [comment]
Hi @aniket02k - have you had a chance to take #889 (or the pushed image) for a spin?   Please update when you have.

**Comment by aniket02k** [comment]
Hi Kevin,
Thanks for the fix :)
I've verified the fix.  When `KERNEL_IMAGE` is being passed from client to eg, the `KERNEL_IMAGE` flows from the client to eg.

Regards,
Aniket

**Comment by kevin-bates** [comment]
Great, thanks for the update.  We should get this merged today and I'm hoping we can have an EG 2.3 release out in the next day or two.
